packages/graphiql/src/state/EditorProvider.tsx

```diff
@@ -0,0 +1,87 @@
+import React from 'react';
+
+import { EditorContexts } from './types';
+import useDispatchedCallback from '../hooks/useDispatchedCallback';
+
+interface EditorState {
+  editor?: CodeMirror.Editor;
+  text?: string;
+}
+const initialState: EditorState = {};
+
+enum EditorActionType {
+  Load = 'load',
+  Change = 'change',
+}
+
+const editorChangeAction = (payload: string) =>
+  ({
+    type: EditorActionType.Change,
+    payload,
+  } as const);
+
+const editorLoadedAction = (payload: CodeMirror.Editor) =>
+  ({
+    type: EditorActionType.Load,
+    payload,
+  } as const);
+
+type EditorAction =
+  | ReturnType<typeof editorChangeAction>
+  | ReturnType<typeof editorLoadedAction>;
+
+function editorReducer(state: EditorState, action: EditorAction): EditorState {
+  switch (action.type) {
+    case EditorActionType.Load: {
+      return {
+        ...state,
+        editor: action.payload,
+      };
+    }
+    case EditorActionType.Change: {
+      return {
+        ...state,
+        text: action.payload,
+      };
+    }
+    default: {
+      return state;
+    }
+  }
+}
+
+type EditorContextType =
+  | (EditorState & {
+      context: EditorContexts;
+      onLoadEditor: (...args: Parameters<typeof editorLoadedAction>) => void;
+      onChange: (...args: Parameters<typeof editorChangeAction>) => void;
```

Comment:
both this and `onLoadEditor` would be required, as in the case of results we have a "controlled component" type situation, where we don't care about `onChange` because it's static. possibly we can have a `controlled={false}` by default for the `EditorProvider`?

Reply:
would a `readOnly` prop cover that you think?